pkg/deployment/agency/state.go

```diff
@@ -0,0 +1,117 @@
+//
+// DISCLAIMER
+//
+// Copyright 2016-2021 ArangoDB GmbH, Cologne, Germany
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+
+package agency
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/arangodb/go-driver"
+	"github.com/arangodb/go-driver/agency"
+	"github.com/arangodb/kube-arangodb/pkg/util/errors"
+)
+
+func loadState(ctx context.Context, client agency.Agency) (State, error) {
+	conn := client.Connection()
+
+	req, err := client.Connection().NewRequest(http.MethodPost, "/_api/agency/read")
+	if err != nil {
+		return State{}, err
+	}
+
+	var data []byte
+
+	req, err = req.SetBody(GetAgencyReadRequest(GetAgencyReadKey(GetAgencyKey(ArangoKey, SupervisionKey, SupervisionMaintenanceKey), GetAgencyKey(ArangoKey, PlanKey, PlanCollectionsKey), GetAgencyKey(ArangoKey, CurrentKey, PlanCollectionsKey))))
+	if err != nil {
+		return State{}, err
+	}
+
+	resp, err := conn.Do(driver.WithRawResponse(ctx, &data), req)
+	if err != nil {
+		return State{}, err
+	}
+
+	if err := resp.CheckStatus(http.StatusOK); err != nil {
+		return State{}, err
+	}
+
+	var c StateRoots
+
+	if err := json.Unmarshal(data, &c); err != nil {
+		return State{}, err
+	}
+
+	if len(c) != 1 {
+		return State{}, errors.Newf("Invalid response size")
+	}
+
+	state := c[0].Arango
+
+	if _, ok := state.Current.Collections["_system"]; !ok {
+		return State{}, errors.Newf("Unable to find system database (invalid data)")
+	}
+
+	if _, ok := state.Plan.Collections["_system"]; !ok {
+		return State{}, errors.Newf("Unable to find system database (invalid data)")
```

Comment:
Unable to find system database (invalid data) -> Unable to find system database in the plan (invalid data)

Reply:
This error should never happen, unless agency is not totally broken